cryptol/REPL/Haskeline.hs

```diff
@@ -223,22 +223,32 @@ canDisplayColor = io (hSupportsANSI stdout)
 cryptolCommand :: CompletionFunc REPL
 cryptolCommand cursor@(l,r)
   | ":" `isPrefixOf` l'
-  , Just (_,cmd,rest) <- splitCommand l' = case nub (findCommand cmd) of
-
-      [c] | null rest && not (any isSpace l') -> do
-            return (l, cmdComp cmd c)
-          | otherwise -> do
-            (rest',cs) <- cmdArgument (cBody c) (reverse (sanitize rest),r)
-            return (unwords [rest', reverse cmd],cs)
-
-      cmds ->
-        return (l, concat [ cmdComp l' c | c <- cmds ])
+  , Just (_,cmd,rest) <- splitCommand l' =
+    case (findCommandExact cmd, findCommand cmd) of
+      -- If we find a single command that matches the input exactly, use that.
+      -- We make this a special case because if the user types :check, then we
+      -- specifically want to complete the :check command, and /not/ the
+      -- :check-docstrings command (for which :check is a prefix). Without this
+      -- special case, there would be ambiguity about which command to complete.
+      ([c], _)  -> completeCommand cmd rest c
+      -- Otherwise, look for commands that are a prefix of the input.
+      (_, [c])  -> completeCommand cmd rest c
+      (_, cmds) -> return (l, concat [ cmdComp l' c | c <- cmds ])
   -- Complete all : commands when the line is just a :
   | ":" == l' = return (l, concat [ cmdComp l' c | c <- nub (findCommand ":") ])
   | otherwise = completeExpr cursor
   where
   l' = sanitize (reverse l)
 
+  completeCommand ::
+    String -> String -> CommandDescr -> REPL (String, [Completion])
+  completeCommand cmd rest c
+    | null rest && not (any isSpace l') = do
+      return (l, cmdComp cmd c)
+    | otherwise = do
+      (rest',cs) <- cmdArgument (cBody c) (reverse (sanitize rest),r)
+      return (unwords [rest', reverse cmd],cs)
+
```

Comment:
This completes ```:check```<tab> to ```:check ``` as well as handling the completions for :check's arguments. So if you type ```:ch<tab>``` it completes as far as ```:check``` but then another tab doesn't tell you what the remaining expansions are but just completes ```:check```. This seems not entirely desirable.

I think it should check ```null rest && not (any isSpace l')``` first and only go on to look at ```findCommandExact``` if that isn't true. (And in that case I think it wants to only look at ```findCommandExact```...)

Reply:
Yes, a fair point. On a related note, it would be nice to add some unit tests for this code, since there are quite a few interesting cases here that are tedious to check manually.

Reply:
I've pushed an additional commit which adds some scaffolding for Cryptol API unit tests, and I've filled out the test cases that work thus far. I haven't gotten around to actually implementing your suggested fix, however. I'm not sure when I will get back around to working on this, so if someone would like to finish this PR in the interim, please feel free to push to the branch.

Reply:
I've pushed a commit which fixes the bugs that @sauclovian-g noticed, and I've added even more regression tests. PTAL.